WalletWasabi/Tor/Socks5/Pool/TorPoolItem.cs

```diff
@@ -0,0 +1,146 @@
+		private object StateLock { get; } = new object();
+
+		/// <remarks>All access to this property must be guarded by <see cref="StateLock"/>.</remarks>
+		public PoolItemState State { get; private set; }
+
+		/// <summary>Tor SOCKS5 connection associated with this pool item.</summary>
+		public TorConnection Client { get; }
+
+		/// <summary>Gets whether this pool item can be potentially re-used.</summary>
+		private bool AllowRecycling { get; }
+
+		/// <summary>Unique identifier of the pool item for logging purposes.</summary>
+		private long Id { get; }
+
+		public Stream GetTransportStream()
+		{
+			lock (StateLock)
+			{
+				return Client.GetTransportStream();
+			}
+		}
+
+		/// <summary>
+		/// Gets whether internal <see cref="TorConnection"/> can be re-used for a new HTTP(s) request.
+		/// </summary>
+		/// <returns><c>true</c> when <see cref="TorPoolItem"/> must be disposed, <c>false</c> otherwise.</returns>
+		public bool NeedRecycling
+		{
+			get
+			{
+				lock (StateLock)
+				{
+					return (State == PoolItemState.ToDispose) || (AllowRecycling && (State == PoolItemState.FreeToUse) && !Client.IsConnected);
+				}
+			}
+		}
+
+		public bool TryReserve()
+		{
+			lock (StateLock)
+			{
+				if (State == PoolItemState.FreeToUse && Client.IsConnected)
+				{
+					State = PoolItemState.InUse;
+					return true;
+				}
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		/// After the <see cref="TorPoolItem"/> is used to send an HTTP(s) request, it needs to be unreserved
+		/// so that the pool item can be used again.
+		/// </summary>
+		/// <returns>Pool item state after unreserve operation.</returns>
+		public PoolItemState Unreserve()
+		{
+			lock (StateLock)
+			{
+				if (State < PoolItemState.Disposed)
+				{
+					State = AllowRecycling ? PoolItemState.FreeToUse : PoolItemState.ToDispose;
+				}
+
+				return State;
+			}
+		}
+
+		public override string? ToString()
+		{
+			return $"PoolItem#{Id}";
+		}
+
+		/// <summary>
+		/// <list type="bullet">
+		/// <item>Unmanaged resources need to be released regardless of the value of the <paramref name="disposing"/> parameter.</item>
+		/// <item>Managed resources need to be released if the value of <paramref name="disposing"/> is <c>true</c>.</item>
+		/// </list>
+		/// </summary>
+		/// <param name="disposing">
+		/// Indicates whether the method call comes from a <see cref="Dispose()"/> method
+		/// (its value is <c>true</c>) or from a finalizer (its value is <c>false</c>).
+		/// </param>
+		protected virtual void Dispose(bool disposing)
+		{
+			if (!_disposedValue)
+			{
+				if (disposing)
+				{
+					Client.Dispose();
+					lock (StateLock)
+					{
+						State = PoolItemState.Disposed;
+					}
+				}
+				_disposedValue = true;
+			}
+		}
+
+		/// <summary>
+		/// Do not change this code.
+		/// </summary>
+		public void Dispose()
+		{
+			// Dispose of unmanaged resources.
+			Dispose(true);
+
+			// Suppress finalization.
+			GC.SuppressFinalize(this);
```

Comment:
This is used to indicate the GC that it shouldn't execute the finalizer assuming there is a finalizer that invokes `Dispose(false)`. In this case the class has no finalizer. it could be a good idea to have one `~TorPoolItem() => Dispose(false);`

Reply:
Well, this is a tricky topic for me. However, I don't want to use .NET's finalizer really as what I would like to achieve here is:

> Every time you remove an `IPoolItem` from `HostBuckets` in `TorPoolItemManager`, I want to explicitly dispose `IPoolItem`.

If this condition is true, then I don't see a way to create a memory leak. 

Do you agree with this or not? Do you find it reasonable?

Reply:
That's ok. In that case i probably wouldn't call the method Dispose because it creates confusion, something like Release could be better. I also would remove the protected virtual Dispose(bool).